File: src/app/services/pricing.service.ts
// Angular core and common modules
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';

// RxJS modules
import { Observable, of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';

// Environment-specific imports
import { environment } from '../../environments/environment';

// Core interfaces for better type safety
export interface RampPricing {
  readonly minSeats: number;
  readonly price: number;
}

export interface Plan {
  readonly name: string;
  readonly description: string;
  readonly features: readonly string[];
  readonly oneYearPricing: readonly RampPricing[];
  readonly threeYearPricing: readonly RampPricing[];
  readonly packagePrice?: number;
  readonly apiPrice?: number;
  readonly freePackagesPerSeat?: number;
}

export interface ProductFamily {
  readonly name: string;
  readonly description: string;
  readonly plans: readonly Plan[];
}

export interface CartItem {
  readonly productFamily: string;
  readonly planName: string;
  readonly seats: number;
  readonly packages?: number;
  readonly apiCalls?: number;
  readonly term: BillingTerm;
  readonly price: number;
}

export interface PlanPricing {
  readonly basePrice: number;
  readonly seatPrice?: number;
  readonly packagePrice?: number;
  readonly apiCallPrice?: number;
}

// Backend API Response Interfaces
export interface PricingApiResponse {
  readonly productFamilies: ProductFamily[];
  readonly supportedCurrencies: readonly string[];
  readonly lastUpdated: string;
}

export interface EstimateItemRequest {
  readonly productFamily: string;
  readonly planName: string;
  readonly seats: number;
  readonly packages?: number;
  readonly apiCalls?: number;
}

export interface EstimateRequest {
  items: EstimateItemRequest[];
  currency: string;
  billingTerm: string;
}

export interface EstimateItemResponse {
  productFamily: string;
  planName: string;
  seats: number;
  basePrice: number;
  packagesPrice: number;
  apiCallsPrice: number;
  totalPrice: number;
  appliedTier: string;
  includedPackages?: number;
  extraPackages?: number;
}

export interface EstimateResponse {
  items: EstimateItemResponse[];
  subtotal: number;
  total: number;
  currency: string;
  billingTerm: string;
}

// Type unions for better type safety
export type BillingTerm = '1year' | '3year';
export type SupportedCurrency = 'USD' | 'EUR' | 'GBP' | 'CAD' | 'AUD';

// Constants for better maintainability
export const BILLING_TERMS = {
  ONE_YEAR: '1year',
  THREE_YEAR: '3year'
} as const;

export const DEFAULT_FREE_PACKAGES_PER_SEAT = 200;

export const SUPPORTED_CURRENCIES: readonly SupportedCurrency[] = [
  'USD',
  'EUR', 
  'GBP',
  'CAD',
  'AUD'
] as const;

@Injectable({
  providedIn: 'root'
})
export class PricingService {
  private readonly fallbackPricingDataUrl = environment.pricingDataUrl;
  private readonly apiUrl = environment.apiUrl;
  private cachedPricingData: PricingApiResponse | null = null;

  constructor(private readonly httpClient: HttpClient) {}

  /**
   * Fetch pricing data from backend API with fallback to static data
   */
  fetchPricingData(currency: string = 'USD'): Observable<PricingApiResponse> {
    console.log(`🔄 Fetching pricing data from backend API for currency: ${currency}`);
    
    return this.httpClient.get<PricingApiResponse>(`${this.apiUrl}/pricing?currency=${currency}`).pipe(
      map(response => {
        console.log('✅ Successfully fetched pricing data from backend API:', {
          productFamilies: response.productFamilies?.length || 0,
          supportedCurrencies: response.supportedCurrencies?.length || 0,
          lastUpdated: response.lastUpdated
        });
        this.cachedPricingData = response;
        return response;
      }),
      catchError(error => {
        console.warn('⚠️ Failed to fetch pricing from backend API, falling back to static data:', error);
        return this.fetchFallbackPricingData();
      })
    );
  }

  /**
   * Fallback to static pricing data when backend is unavailable
   */
  private fetchFallbackPricingData(): Observable<PricingApiResponse> {
    console.log('📁 Falling back to static pricing data from:', this.fallbackPricingDataUrl);
    
    return this.httpClient.get<{ productFamilies: ProductFamily[] }>(this.fallbackPricingDataUrl).pipe(
      map(staticData => {
        console.log('✅ Successfully loaded fallback static data:', {
          productFamilies: staticData.productFamilies?.length || 0
        });
        return {
          productFamilies: staticData.productFamilies || [],
          supportedCurrencies: ['USD', 'EUR', 'GBP', 'CAD', 'AUD'],
          lastUpdated: new Date().toISOString()
        };
      }),
      catchError(error => {
        console.error('❌ Failed to load fallback static data:', error);
        return of({
          productFamilies: [],
          supportedCurrencies: ['USD'],
          lastUpdated: new Date().toISOString()
        });
      })
    );
  }

  /**
   * Get pricing estimate from backend API with fallback to local calculation
   */
  getEstimate(request: EstimateRequest): Observable<EstimateResponse> {
    console.log('🧮 Requesting estimate from backend API:', {
      items: request.items.length,
      currency: request.currency,
      billingTerm: request.billingTerm
    });
    
    return this.httpClient.post<EstimateResponse>(`${this.apiUrl}/estimate`, request).pipe(
      map(response => {
        console.log('✅ Successfully received estimate from backend API:', {
          subtotal: response.subtotal,
          total: response.total,
          itemsCount: response.items.length
        });
        return response;
      }),
      catchError(error => {
        console.warn('⚠️ Failed to get estimate from backend API, falling back to local calculation:', error);
        return this.calculateLocalEstimate(request);
      })
    );
  }

  /**
   * Fallback estimate calculation using cached pricing data
   */
  private calculateLocalEstimate(request: EstimateRequest): Observable<EstimateResponse> {
    if (!this.cachedPricingData) {
      return of({
        items: [],
        subtotal: 0,
        total: 0,
        currency: request.currency,
        billingTerm: request.billingTerm
      });
    }

    const items = request.items.map(item => this.calculateItemEstimate(item, request.billingTerm));
    const subtotal = items.reduce((sum, item) => sum + item.totalPrice, 0);

    return of({
      items,
      subtotal,
      total: subtotal,
      currency: request.currency,
      billingTerm: request.billingTerm
    });
  }

  /**
   * Calculate estimate for a single item using cached pricing data
   */
  private calculateItemEstimate(item: EstimateItemRequest, billingTerm: string): EstimateItemResponse {
    const family = this.cachedPricingData?.productFamilies.find(f => f.name === item.productFamily);
    const plan = family?.plans.find(p => p.name === item.planName);

    if (!plan) {
      return {
        productFamily: item.productFamily,
        planName: item.planName,
        seats: item.seats,
        basePrice: 0,
        packagesPrice: 0,
        apiCallsPrice: 0,
        totalPrice: 0,
        appliedTier: 'Unknown'
      };
    }

    const pricing = billingTerm.toLowerCase().includes('3') ? plan.threeYearPricing : plan.oneYearPricing;
    const applicableTier = [...pricing]
      .sort((a, b) => b.minSeats - a.minSeats)
      .find(tier => item.seats >= tier.minSeats);

    const basePrice = applicableTier ? item.seats * applicableTier.price : 0;
    const appliedTier = applicableTier ? `$${applicableTier.price}/seat` : 'Unknown';

    // Calculate packages cost
    let packagesPrice = 0;
    let includedPackages: number | undefined;
    let extraPackages: number | undefined;

    if (item.packages && plan.packagePrice) {
      if (plan.freePackagesPerSeat) {
        includedPackages = item.seats * plan.freePackagesPerSeat;
        extraPackages = Math.max(0, item.packages - includedPackages);
        packagesPrice = extraPackages * plan.packagePrice;
      } else {
        packagesPrice = item.packages * plan.packagePrice;
        extraPackages = item.packages;
      }
    }

    // Calculate API calls cost
    const apiCallsPrice = (item.apiCalls || 0) * (plan.apiPrice || 0);

    return {
      productFamily: item.productFamily,
      planName: item.planName,
      seats: item.seats,
      basePrice,
      packagesPrice,
      apiCallsPrice,
      totalPrice: basePrice + packagesPrice + apiCallsPrice,
      appliedTier,
      includedPackages,
      extraPackages
    };
  }

  /**
   * Pure function - calculate price locally (fallback method)
   */
  calculatePrice(
    productFamily: ProductFamily, 
    planName: string, 
    seats: number, 
    packages: number = 0, 
    apiCalls: number = 0, 
    term: BillingTerm
  ): number {
    const plan = this.findPlanByName(productFamily, planName);
    if (!plan) {
      return 0;
    }

    const pricing = term === BILLING_TERMS.ONE_YEAR ? plan.oneYearPricing : plan.threeYearPricing;
    const applicableTier = this.findApplicablePricingTier(pricing, seats);

    if (!applicableTier) {
      return 0;
    }

    let total = seats * applicableTier.price;

    if (plan.freePackagesPerSeat && plan.packagePrice) {
      const freePackages = seats * plan.freePackagesPerSeat;
      const extraPackages = Math.max(0, packages - freePackages);
      total += extraPackages * plan.packagePrice;
    }

    if (plan.apiPrice && apiCalls > 0) {
      total += apiCalls * plan.apiPrice;
    }

    return total;
  }

  /**
   * Pure function - find plan by name
   */
  private findPlanByName(productFamily: ProductFamily, planName: string): Plan | null {
    return productFamily.plans.find(plan => plan.name === planName) ?? null;
  }

  /**
   * Pure function - find applicable pricing tier
   */
  private findApplicablePricingTier(pricing: readonly RampPricing[], seats: number): RampPricing | null {
    return [...pricing]
      .sort((a, b) => b.minSeats - a.minSeats)
      .find(tier => seats >= tier.minSeats) ?? null;
  }

  /**
   * Legacy method for backward compatibility - get ramp price per seat
   */
  getRampPrice(plan: any, seats: number, term: '1year' | '3year'): number {
    const pricing = term === '1year' ? plan.oneYearPricing : plan.threeYearPricing;
    const applicableTier = [...pricing]
      .sort((a, b) => b.minSeats - a.minSeats)
      .find(tier => seats >= tier.minSeats);
    return applicableTier ? applicableTier.price : 0;
  }

  /**
   * Get supported currencies from backend or default list
   */
  getSupportedCurrencies(): readonly SupportedCurrency[] {
    return this.cachedPricingData?.supportedCurrencies as readonly SupportedCurrency[] ?? SUPPORTED_CURRENCIES;
  }

  /**
   * Check if backend API is available
   */
  checkApiHealth(): Observable<boolean> {
    return this.httpClient.get(`${this.apiUrl}/health`).pipe(
      map(() => true),
      catchError(() => of(false))
    );
  }
}

File: backend/src/main/scala/com/nitro/pricing/services/PricingService.scala
package com.nitro.pricing.services

import com.nitro.pricing.models.{ProductStructure, ChargebeeItem, ChargebeeItemPrice, RampPrice, PricingPlan, PricingProductFamily, PricingApiResponse, PricingEstimateRequest, PricingEstimateResponse, PricingEstimateItemRequest, PricingEstimateItemResponse}
import com.typesafe.scalalogging.LazyLogging
import io.circe.parser.decode
import io.circe.generic.auto._

import scala.concurrent.{ExecutionContext, Future}
import scala.io.Source
import scala.util.{Try, Success, Failure}

case class StaticPricingData(
  productFamilies: List[PricingProductFamily]
)

class PricingService(
  chargebeeClient: ChargebeeClient
)(implicit ec: ExecutionContext) extends LazyLogging {

  private val staticPricingFile = "pricing-data.json"
  
  // Cache for static pricing data (3-year pricing)
  private var cachedStaticData: Option[StaticPricingData] = None
  
  // Cache for Chargebee data with TTL
  private var cachedChargebeeData: Option[(ProductStructure, Long)] = None
  private val chargebeeCacheTtlMs = 60 * 60 * 1000 // 1 hour TTL
  
  def getPricing(currency: String = "USD"): Future[PricingApiResponse] = {
    logger.info(s"Fetching pricing data for currency: $currency")
    
    for {
      staticData <- loadStaticPricingData()
      chargebeeData <- getCachedChargebeeData()
    } yield {
      val mergedProductFamilies = mergeChargebeeWithStaticData(staticData, chargebeeData, currency)
      
      PricingApiResponse(
        productFamilies = mergedProductFamilies,
        supportedCurrencies = List("USD", "EUR", "GBP", "CAD", "AUD"),
        lastUpdated = java.time.Instant.now().toString
      )
    }
  }
  
  def calculateEstimate(request: PricingEstimateRequest): Future[PricingEstimateResponse] = {
    logger.info(s"Calculating estimate for ${request.items.length} items, currency: ${request.currency}, term: ${request.billingTerm}")
    
    for {
      pricingData <- getPricing(request.currency)
    } yield {
      val itemResponses = request.items.map { item =>
        calculateItemEstimate(item, pricingData, request.billingTerm)
      }
      
      val subtotal = itemResponses.map(_.totalPrice).sum
      
      PricingEstimateResponse(
        items = itemResponses,
        subtotal = subtotal,
        total = subtotal, // Tax calculation will be added later
        currency = request.currency,
        billingTerm = request.billingTerm
      )
    }
  }
  
  private def loadStaticPricingData(): Future[StaticPricingData] = {
    cachedStaticData match {
      case Some(data) => Future.successful(data)
      case None =>
        Future {
          val source = Source.fromResource(staticPricingFile)
          val jsonString = try {
            source.mkString
          } finally {
            source.close()
          }
          
          decode[StaticPricingData](jsonString) match {
            case Right(data) =>
              logger.info("Successfully loaded static pricing data")
              cachedStaticData = Some(data)
              data
            case Left(error) =>
              logger.error(s"Failed to parse static pricing data: $error")
              throw new RuntimeException(s"Failed to load static pricing data: $error")
          }
        }
    }
  }
  
  private def getCachedChargebeeData(): Future[ProductStructure] = {
    val currentTime = System.currentTimeMillis()
    
    cachedChargebeeData match {
      case Some((data, cachedTime)) if (currentTime - cachedTime) < chargebeeCacheTtlMs =>
        logger.debug("Using cached Chargebee data")
        Future.successful(data)
      case _ =>
        logger.info("Fetching fresh Chargebee data")
        chargebeeClient.getProductStructure().map { data =>
          cachedChargebeeData = Some((data, currentTime))
          data
        }.recover {
          case ex =>
            logger.warn(s"Failed to fetch Chargebee data, using cached if available: ${ex.getMessage}")
            cachedChargebeeData.map(_._1).getOrElse(throw ex)
        }
    }
  }
  
  private def mergeChargebeeWithStaticData(
    staticData: StaticPricingData,
    chargebeeData: ProductStructure,
    currency: String
  ): List[PricingProductFamily] = {
    
    logger.info(s"Merging pricing data: ${staticData.productFamilies.length} static families, ${chargebeeData.items.length} Chargebee items")
    
    // Map to store Chargebee pricing by item name
    val chargebeePricingMap = chargebeeData.itemPrices
      .filter(_.currency_code.toUpperCase == currency.toUpperCase)
      .groupBy(_.item_id)
    
    staticData.productFamilies.map { family =>
      val updatedPlans = family.plans.map { plan =>
        // Try to find corresponding Chargebee pricing for 1-year terms
        val chargebeeOneYearPricing = findChargebeePricingForPlan(plan.name, chargebeePricingMap)
        
        plan.copy(
          oneYearPricing = chargebeeOneYearPricing.getOrElse(plan.oneYearPricing)
          // Keep threeYearPricing from static data unchanged
        )
      }
      
      family.copy(plans = updatedPlans)
    }
  }
  
  private def findChargebeePricingForPlan(
    planName: String,
    chargebeePricingMap: Map[String, List[ChargebeeItemPrice]]
  ): Option[List[RampPrice]] = {
    
    // Map plan names to Chargebee item IDs
    val itemId = planName match {
      case "Nitro PDF Standard" => "Nitro_PDF_STD"
      case "Nitro PDF Plus" => "Nitro_PDF_PLUS"
      case "Nitro Sign Standard" => "Nitro_SIGN_STD"
      case "Nitro Sign Plus" => "Nitro_SIGN_PLUS"
      case "Nitro Sign Enterprise" => "Nitro_SIGN_ENT"
      case _ => 
        logger.warn(s"No Chargebee mapping found for plan: $planName")
        return None
    }
    
    logger.debug(s"Looking for Chargebee pricing for plan '$planName' -> item '$itemId'")
    
    chargebeePricingMap.get(itemId) match {
      case Some(itemPrices) =>
        logger.debug(s"Found ${itemPrices.length} item prices for $itemId")
        // Find yearly pricing (period = 1, period_unit = "year")
        itemPrices.find(ip => ip.period == 1 && ip.period_unit == "year") match {
          case Some(itemPrice) =>
            logger.debug(s"Found yearly pricing for $itemId: ${itemPrice.id}")
            itemPrice.tiers match {
              case Some(tierList) =>
                logger.debug(s"Found ${tierList.length} tiers for $itemId")
                val rampPrices = tierList.map { tier =>
                  RampPrice(
                    minSeats = tier.starting_unit,
                    price = BigDecimal(tier.price_in_decimal)
                  )
                }.sortBy(_.minSeats) // Ensure proper ordering
                logger.debug(s"Converted tiers for $itemId: ${rampPrices.take(2)}")
                Some(rampPrices)
              case None =>
                logger.warn(s"No tiers found for $itemId pricing")
                None
            }
          case None =>
            logger.warn(s"No yearly pricing found for $itemId (available periods: ${itemPrices.map(ip => s"${ip.period} ${ip.period_unit}").mkString(", ")})")
            None
        }
      case None =>
        logger.warn(s"No item prices found for $itemId. Available items: ${chargebeePricingMap.keys.mkString(", ")}")
        None
    }
  }
  
  private def calculateItemEstimate(
    item: PricingEstimateItemRequest,
    pricingData: PricingApiResponse,
    billingTerm: String
  ): PricingEstimateItemResponse = {
    
    // Find the product family and plan
    val productFamily = pricingData.productFamilies.find(_.name == item.productFamily)
      .getOrElse(throw new IllegalArgumentException(s"Product family not found: ${item.productFamily}"))
    
    val plan = productFamily.plans.find(_.name == item.planName)
      .getOrElse(throw new IllegalArgumentException(s"Plan not found: ${item.planName}"))
    
    // Select pricing based on billing term
    val pricing = billingTerm match {
      case "1year" => plan.oneYearPricing
      case "3year" => plan.threeYearPricing
      case _ => throw new IllegalArgumentException(s"Invalid billing term: $billingTerm")
    }
    
    // Find applicable tier
    val appliedTier = pricing
      .filter(_.minSeats <= item.seats)
      .maxByOption(_.minSeats)
      .getOrElse(throw new IllegalArgumentException(s"No pricing tier found for ${item.seats} seats"))
    
    // Calculate base price
    val basePrice = appliedTier.price * item.seats
    
    // Calculate packages price
    val (packagesPrice, includedPackages, extraPackages) = plan.freePackagesPerSeat match {
      case Some(freePerSeat) =>
        val totalIncluded = freePerSeat * item.seats
        val requestedPackages = item.packages.getOrElse(0)
        val extra = math.max(0, requestedPackages - totalIncluded)
        val packageCost = extra * plan.packagePrice.getOrElse(BigDecimal(0))
        (packageCost, Some(totalIncluded), Some(extra))
      case None =>
        val packageCost = item.packages.getOrElse(0) * plan.packagePrice.getOrElse(BigDecimal(0))
        (packageCost, None, item.packages)
    }
    
    // Calculate API calls price
    val apiCallsPrice = item.apiCalls.getOrElse(0) * plan.apiPrice.getOrElse(BigDecimal(0))
    
    // Total price for this item
    val totalPrice = basePrice + packagesPrice + apiCallsPrice
    
    PricingEstimateItemResponse(
      productFamily = item.productFamily,
      planName = item.planName,
      seats = item.seats,
      basePrice = basePrice,
      packagesPrice = packagesPrice,
      apiCallsPrice = apiCallsPrice,
      totalPrice = totalPrice,
      appliedTier = appliedTier,
      includedPackages = includedPackages,
      extraPackages = extraPackages
    )
  }
}
